api/hub/service.go

```diff
@@ -0,0 +1,468 @@
+package hub
+
+import (
+	"context"
+	"fmt"
+	"net/mail"
+	"time"
+
+	logging "github.com/ipfs/go-log"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	threads "github.com/textileio/go-threads/api/client"
+	"github.com/textileio/go-threads/broadcast"
+	"github.com/textileio/go-threads/core/thread"
+	"github.com/textileio/textile/api/common"
+	pb "github.com/textileio/textile/api/hub/pb"
+	c "github.com/textileio/textile/collections"
+	"github.com/textileio/textile/email"
+	"github.com/textileio/textile/util"
+	"go.mongodb.org/mongo-driver/mongo"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	log = logging.Logger("hub")
+
+	loginTimeout = time.Minute * 3
+	emailTimeout = time.Second * 10
+)
+
+type Service struct {
+	Collections *c.Collections
+	Threads     *threads.Client
+
+	EmailClient *email.Client
+
+	GatewayUrl string
+
+	SessionBus    *broadcast.Broadcaster
+	SessionSecret string
+}
+
+func (s *Service) Signup(ctx context.Context, req *pb.SignupRequest) (*pb.SignupReply, error) {
+	log.Debugf("received signup request")
+
+	if err := s.Collections.Accounts.ValidateUsername(req.Username); err != nil {
+		return nil, status.Error(codes.FailedPrecondition, err.Error())
+	}
+	if _, err := mail.ParseAddress(req.Email); err != nil {
+		return nil, status.Error(codes.FailedPrecondition, "Email address in not valid")
+	}
+
+	secret := getSessionSecret(s.SessionSecret)
+	ectx, cancel := context.WithTimeout(ctx, emailTimeout)
+	defer cancel()
+	if err := s.EmailClient.ConfirmAddress(ectx, req.Email, s.GatewayUrl, secret); err != nil {
+		return nil, err
+	}
+	if !s.awaitVerification(secret) {
+		return nil, status.Error(codes.Unauthenticated, "Could not verify email address")
+	}
+
+	dev, err := s.Collections.Accounts.CreateDev(ctx, req.Username, req.Email)
+	if err != nil {
+		return nil, status.Error(codes.FailedPrecondition, "Account exists")
+	}
+	session, err := s.Collections.Sessions.Create(ctx, dev.Key)
+	if err != nil {
+		return nil, err
+	}
+	ctx = common.NewSessionContext(ctx, session.ID)
+	tok, err := s.Threads.GetToken(ctx, thread.NewLibp2pIdentity(dev.Secret))
+	if err != nil {
+		return nil, err
+	}
+	if err := s.Collections.Accounts.SetToken(ctx, dev.Key, tok); err != nil {
+		return nil, err
+	}
+
+	// Check for pending invites
+	invites, err := s.Collections.Invites.List(ctx, dev.Email)
+	if err != nil {
+		return nil, err
+	}
+	for _, invite := range invites {
+		if invite.Accepted {
```

Comment:
If an existing user clicks the accept invite link in an email, then they get auto-added to the org and the invite object is deleted. However, if the person clicking the accept link does not have an account, we don't auto create their account, just mark the invite as "accepted". Then, when they do go to create an account, the accepted invite is found, at which point they are added to the org and the invite is deleted.

Reply:
I see. So anyone with the link can accept the invite. Hopefully, that _anyone_ is only the intended person. Never thought about that workflow for invites... but I guess it's safe since whoever accepts the invite is controlling the email, so sounds good.

Reply:
Yep, but maybe eventually when we have a dashboard with signin, etc., you'd be forced to login or signup before the invite is processed.